File: ios/Classes/Audio/MinisAudioPlugin.swift
import AVFoundation
import Flutter
import Foundation

/// Hosts the `loopit/minis/audio` channels and dispatches into the session /
/// player / recorder / waveform extractor singletons.
public final class MinisAudioPlugin: NSObject {
    private let session = MinisAudioSession()
    private let players = MinisPlayerRegistry()
    private let recorder = MinisRecorder()
    private let waveform = MinisWaveformExtractor()
    private let routes = MinisAudioRouteWatcher()
    private let trimmer = MinisAudioTrimmer()
    private let mixer = MinisAudioMixer()
    private let normalizer = MinisLufsNormalizer()
    private let beats = MinisBeatDetector()
    private let pitcher = MinisPitchTimeStretch()

    private let methodChannel: FlutterMethodChannel
    private let playerEventsChannel: FlutterEventChannel
    private let levelsChannel: FlutterEventChannel
    private let stateChannel: FlutterEventChannel
    private let progressChannel: FlutterEventChannel

    private let playerEventsHandler = MinisStreamHandler()
    private let levelsHandler = MinisStreamHandler()
    private let stateHandler = MinisStreamHandler()
    private let progressHandler = MinisStreamHandler()
    private var progressSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let plugin = MinisAudioPlugin(messenger: registrar.messenger())
        registrar.publish(plugin)
    }

    init(messenger: FlutterBinaryMessenger) {
        methodChannel = FlutterMethodChannel(
            name: "loopit/minis/audio",
            binaryMessenger: messenger
        )
        playerEventsChannel = FlutterEventChannel(
            name: "loopit/minis/audio/playerEvents",
            binaryMessenger: messenger
        )
        levelsChannel = FlutterEventChannel(
            name: "loopit/minis/audio/levels",
            binaryMessenger: messenger
        )
        stateChannel = FlutterEventChannel(
            name: "loopit/minis/audio/state",
            binaryMessenger: messenger
        )
        progressChannel = FlutterEventChannel(
            name: "loopit/minis/audio/progress",
            binaryMessenger: messenger
        )
        super.init()
        playerEventsHandler.onSinkChanged = { [weak self] sink in
            self?.players.attachSink(sink)
        }
        levelsHandler.onSinkChanged = { [weak self] sink in
            self?.recorder.attachLevelSink(sink)
        }
        stateHandler.onSinkChanged = { [weak self] sink in
            guard let self = self else { return }
            if sink != nil {
                self.routes.attachSink(sink)
                self.routes.start()
            } else {
                self.routes.stop()
                self.routes.attachSink(nil)
            }
        }
        progressHandler.onSinkChanged = { [weak self] sink in
            self?.progressSink = sink
        }
        playerEventsChannel.setStreamHandler(playerEventsHandler)
        levelsChannel.setStreamHandler(levelsHandler)
        stateChannel.setStreamHandler(stateHandler)
        progressChannel.setStreamHandler(progressHandler)
        methodChannel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call: call, result: result)
        }
    }

    private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "configureSession":
            let accepted = session.configure(args: args)
            result(["accepted": accepted])
        case "setSessionActive":
            let active = args["active"] as? Bool ?? false
            let applied = session.setActive(active)
            result(["active": applied])
        case "playerCreate":
            guard let path = args["path"] as? String else {
                result(FlutterError(code: "ARG", message: "path required", details: nil))
                return
            }
            let volume = (args["volume"] as? NSNumber)?.floatValue ?? 1.0
            let existingId = (args["existingId"] as? NSNumber)?.intValue
            do {
                let out = try players.createOrReplace(
                    path: path, volume: volume, existingId: existingId
                )
                result(out)
            } catch {
                result(FlutterError(code: "MINIS_AUDIO", message: error.localizedDescription, details: nil))
            }
        case "playerControl":
            guard let id = (args["playerId"] as? NSNumber)?.intValue,
                  let op = args["op"] as? String else {
                result(FlutterError(code: "ARG", message: "playerId+op required", details: nil)); return
            }
            players.control(id: id, op: op, value: args["value"])
            result(nil)
        case "playerSetFinishMode":
            guard let id = (args["playerId"] as? NSNumber)?.intValue else {
                result(FlutterError(code: "ARG", message: "playerId required", details: nil)); return
            }
            players.setFinishMode(id: id, mode: args["mode"] as? String ?? "pause")
            result(nil)
        case "playerGetDuration":
            guard let id = (args["playerId"] as? NSNumber)?.intValue else {
                result(FlutterError(code: "ARG", message: "playerId required", details: nil)); return
            }
            result(["durationMs": players.duration(id: id)])
        case "playerDispose":
            if let id = (args["playerId"] as? NSNumber)?.intValue {
                players.dispose(id: id)
            }
            result(nil)
        case "startRecord":
            guard let path = args["path"] as? String else {
                result(FlutterError(code: "ARG", message: "path required", details: nil)); return
            }
            do {
                try recorder.start(
                    path: path,
                    format: args["format"] as? String ?? "aac",
                    sampleRate: (args["sampleRate"] as? NSNumber)?.intValue ?? 44100,
                    channels: (args["channels"] as? NSNumber)?.intValue ?? 1,
                    bitRate: (args["bitRate"] as? NSNumber)?.intValue ?? 128000,
                    denoise: (args["denoise"] as? Bool) ?? false,
                    monitor: (args["monitor"] as? Bool) ?? false
                )
                result(nil)
            } catch {
                result(FlutterError(code: "MINIS_AUDIO", message: error.localizedDescription, details: nil))
            }
        case "pauseRecord":
            recorder.pause(); result(nil)
        case "resumeRecord":
            recorder.resume(); result(nil)
        case "stopRecord":
            result(recorder.stop())
        case "extractWaveform":
            guard let path = args["path"] as? String else {
                result(FlutterError(code: "ARG", message: "path required", details: nil)); return
            }
            let peaks = (args["peaks"] as? NSNumber)?.intValue ?? 1024
            waveform.extract(path: path, peakCount: peaks, result: result)
        // ---- improvement4.md deferred surface (D1.x/D6/D7/D8/D9) ----
        case "listInputs":
            result(["inputs": session.listInputs()])
        case "setInput":
            guard let id = args["id"] as? String else {
                result(FlutterError(code: "ARG", message: "id required", details: nil)); return
            }
            let ok = session.setPreferredInput(id: id)
            result(["applied": ok])
        case "trim":
            guard let path = args["path"] as? String,
                  let outPath = args["outPath"] as? String else {
                result(FlutterError(code: "ARG", message: "path+outPath required", details: nil)); return
            }
            let inMs = (args["inMs"] as? NSNumber)?.intValue ?? 0
            let outMs = (args["outMs"] as? NSNumber)?.intValue ?? 0
            let mode = args["mode"] as? String ?? "accurate"
            trimmer.trim(path: path, inMs: inMs, outMs: outMs, outPath: outPath, mode: mode, result: result)
        case "mix":
            guard let raw = args["tracks"] as? [[String: Any]],
                  let outPath = args["outPath"] as? String else {
                result(FlutterError(code: "ARG", message: "tracks+outPath required", details: nil)); return
            }
            var parsed: [MinisMixTrack] = []
            parsed.reserveCapacity(raw.count)
            for t in raw {
                guard let p = t["path"] as? String else { continue }
                let inMs = (t["inMs"] as? NSNumber)?.intValue ?? 0
                let outMs = (t["outMs"] as? NSNumber)?.intValue ?? inMs
                let posMs = (t["positionMs"] as? NSNumber)?.intValue ?? 0
                let env = parsePairs(t["gainEnv"])
                let pan = parsePairs(t["panEnv"])
                let eq = t["eq"] as? [String: Any]
                parsed.append(MinisMixTrack(
                    path: p, inMs: inMs, outMs: outMs, positionMs: posMs,
                    gainEnv: env, panEnv: pan,
                    eqLowDb: (eq?["lowDb"] as? NSNumber)?.doubleValue ?? 0,
                    eqMidDb: (eq?["midDb"] as? NSNumber)?.doubleValue ?? 0,
                    eqHighDb: (eq?["highDb"] as? NSNumber)?.doubleValue ?? 0,
                    fadeInMs: (t["fadeInMs"] as? NSNumber)?.doubleValue ?? 0,
                    fadeOutMs: (t["fadeOutMs"] as? NSNumber)?.doubleValue ?? 0,
                    fadeKind: (t["fadeKind"] as? String) ?? "linear",
                    isVoiceForDuck: (t["isVoice"] as? Bool) ?? false
                ))
            }
            let targetLufs = (args["targetLufs"] as? NSNumber)?.doubleValue ?? -14.0
            mixer.mix(tracks: parsed, outPath: outPath, targetLufs: targetLufs, result: result)
        case "normalize":
            guard let path = args["path"] as? String,
                  let outPath = args["outPath"] as? String else {
                result(FlutterError(code: "ARG", message: "path+outPath required", details: nil)); return
            }
            let targetLufs = (args["targetLufs"] as? NSNumber)?.doubleValue ?? -14.0
            normalizer.normalize(path: path, outPath: outPath, targetLufs: targetLufs, result: result)
        case "detectBeats":
            guard let path = args["path"] as? String else {
                result(FlutterError(code: "ARG", message: "path required", details: nil)); return
            }
            beats.detect(path: path, result: result)
        case "pitchShift":
            guard let path = args["path"] as? String,
                  let outPath = args["outPath"] as? String else {
                result(FlutterError(code: "ARG", message: "path+outPath required", details: nil)); return
            }
            let semitones = (args["semitones"] as? NSNumber)?.doubleValue ?? 0.0
            pitcher.pitchShift(path: path, semitones: semitones, outPath: outPath, result: result)
        case "timeStretch":
            guard let path = args["path"] as? String,
                  let outPath = args["outPath"] as? String else {
                result(FlutterError(code: "ARG", message: "path+outPath required", details: nil)); return
            }
            let factor = (args["factor"] as? NSNumber)?.doubleValue ?? 1.0
            let keepPitch = (args["keepPitch"] as? NSNumber)?.boolValue ?? true
            pitcher.timeStretch(
                path: path, factor: factor, keepPitch: keepPitch,
                outPath: outPath, result: result
            )
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func parsePairs(_ raw: Any?) -> [(Double, Double)] {
        var out: [(Double, Double)] = []
        if let arr = raw as? [[NSNumber]] {
            for pair in arr where pair.count >= 2 {
                out.append((pair[0].doubleValue, pair[1].doubleValue))
            }
        } else if let arr = raw as? [[String: Any]] {
            for m in arr {
                if let t = (m["timeMs"] as? NSNumber)?.doubleValue,
                   let g = (m["gain"] ?? m["pan"]) as? NSNumber {
                    out.append((t, g.doubleValue))
                }
            }
        }
        return out
    }

    private func notImpl(_ result: @escaping FlutterResult, _ method: String, _ anchor: String) {
        result(FlutterError(
            code: "NOT_IMPLEMENTED",
            message: "\(method) deferred (see improvement4.md \(anchor))",
            details: nil
        ))
    }
}

final class MinisStreamHandler: NSObject, FlutterStreamHandler {
    var onSinkChanged: ((FlutterEventSink?) -> Void)?

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        onSinkChanged?(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        onSinkChanged?(nil)
        return nil
    }
}

File: android/src/main/kotlin/com/loopit/minis/audio/MinisAudioPlugin.kt
package com.loopit.minis.audio

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Native-side host for the `loopit/minis/audio` channel. Owns the singletons
 * for session, players, recorder, and waveform extraction so the Dart shim can
 * stay stateless.
 */
class MinisAudioPlugin(
    private val context: Context,
    messenger: BinaryMessenger,
) : MethodChannel.MethodCallHandler {

    private val methodChannel = MethodChannel(messenger, "loopit/minis/audio")
    private val playerEvents = EventChannel(messenger, "loopit/minis/audio/playerEvents")
    private val levelsChannel = EventChannel(messenger, "loopit/minis/audio/levels")
    private val stateChannel = EventChannel(messenger, "loopit/minis/audio/state")
    private val progressChannel = EventChannel(messenger, "loopit/minis/audio/progress")

    private val session = AudioSession(context)
    private val players = PlayerRegistry()
    private val recorder = Recorder(context)
    private val waveform = WaveformExtractor(context)
    private val routes = AudioRouteWatcher(context)
    private val trimmer = AudioTrimmer()
    private val mixer = AudioMixer()
    private val normalizer = LufsNormalizer()
    private val beats = BeatDetector()
    private val stretch = PitchTimeStretch()
    private var progressSink: EventChannel.EventSink? = null

    init {
        methodChannel.setMethodCallHandler(this)
        playerEvents.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                players.attachSink(events)
            }

            override fun onCancel(arguments: Any?) {
                players.attachSink(null)
            }
        })
        levelsChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                recorder.attachLevelSink(events)
            }

            override fun onCancel(arguments: Any?) {
                recorder.attachLevelSink(null)
            }
        })
        stateChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                routes.attachSink(events)
                routes.start()
            }

            override fun onCancel(arguments: Any?) {
                routes.stop()
                routes.attachSink(null)
            }
        })
        progressChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                progressSink = events
            }

            override fun onCancel(arguments: Any?) {
                progressSink = null
            }
        })
    }

    fun dispose() {
        methodChannel.setMethodCallHandler(null)
        playerEvents.setStreamHandler(null)
        levelsChannel.setStreamHandler(null)
        stateChannel.setStreamHandler(null)
        progressChannel.setStreamHandler(null)
        routes.stop()
        routes.attachSink(null)
        players.disposeAll()
        recorder.dispose()
        session.release()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "configureSession" -> {
                    val accepted = session.configure(
                        category = call.argument("category"),
                        categoryOptions = (call.argument<Number>("categoryOptions"))?.toInt() ?: 0,
                        mode = call.argument("mode"),
                        androidAttrs = call.argument("androidAttrs"),
                        androidFocusGain = call.argument("androidFocusGain"),
                    )
                    result.success(mapOf("accepted" to accepted))
                }
                "setSessionActive" -> {
                    val active = call.argument<Boolean>("active") ?: false
                    val applied = session.setActive(active)
                    result.success(mapOf("active" to applied))
                }
                "playerCreate" -> {
                    val path = call.argument<String>("path")
                    if (path.isNullOrEmpty()) {
                        result.error("ARG", "path required", null); return
                    }
                    val volume = (call.argument<Number>("volume"))?.toFloat() ?: 1.0f
                    val existingId = (call.argument<Number>("existingId"))?.toInt()
                    val out = players.createOrReplace(path, volume, existingId)
                    result.success(out)
                }
                "playerControl" -> {
                    val id = (call.argument<Number>("playerId"))?.toInt()
                    val op = call.argument<String>("op")
                    if (id == null || op == null) {
                        result.error("ARG", "playerId+op required", null); return
                    }
                    players.control(id, op, call.argument("value"))
                    result.success(null)
                }
                "playerSetFinishMode" -> {
                    val id = (call.argument<Number>("playerId"))?.toInt()
                    val mode = call.argument<String>("mode") ?: "pause"
                    if (id == null) {
                        result.error("ARG", "playerId required", null); return
                    }
                    players.setFinishMode(id, mode)
                    result.success(null)
                }
                "playerGetDuration" -> {
                    val id = (call.argument<Number>("playerId"))?.toInt()
                    if (id == null) {
                        result.error("ARG", "playerId required", null); return
                    }
                    result.success(mapOf("durationMs" to players.duration(id)))
                }
                "playerDispose" -> {
                    val id = (call.argument<Number>("playerId"))?.toInt()
                    if (id != null) players.dispose(id)
                    result.success(null)
                }
                "startRecord" -> {
                    val path = call.argument<String>("path")
                    if (path.isNullOrEmpty()) {
                        result.error("ARG", "path required", null); return
                    }
                    recorder.start(
                        path = path,
                        format = call.argument<String>("format") ?: "aac",
                        sampleRate = (call.argument<Number>("sampleRate"))?.toInt() ?: 44100,
                        channels = (call.argument<Number>("channels"))?.toInt() ?: 1,
                        bitRate = (call.argument<Number>("bitRate"))?.toInt() ?: 128000,
                        denoise = call.argument<Boolean>("denoise") ?: false,
                        monitor = call.argument<Boolean>("monitor") ?: false,
                    )
                    result.success(null)
                }
                "pauseRecord" -> {
                    recorder.pause()
                    result.success(null)
                }
                "resumeRecord" -> {
                    recorder.resume()
                    result.success(null)
                }
                "stopRecord" -> {
                    val out = recorder.stop()
                    result.success(out)
                }
                "extractWaveform" -> {
                    val path = call.argument<String>("path")
                    val peaks = (call.argument<Number>("peaks"))?.toInt() ?: 1024
                    if (path.isNullOrEmpty()) {
                        result.error("ARG", "path required", null); return
                    }
                    waveform.extract(path, peaks, result)
                }
                // ---- improvement4.md deferred surface (D1.x/D6/D7/D8/D9) ----
                // Wired so call sites can probe via PlatformException(code=NOT_IMPLEMENTED).
                "listInputs" -> result.success(mapOf("inputs" to recorder.listInputs()))
                "setInput" -> {
                    val id = call.argument<String>("id")
                    if (id.isNullOrEmpty()) {
                        result.error("ARG", "id required", null); return
                    }
                    val ok = recorder.setPreferredInput(id)
                    result.success(mapOf("applied" to ok))
                }
                "trim" -> {
                    val path = call.argument<String>("path")
                    val outPath = call.argument<String>("outPath")
                    val inMs = (call.argument<Number>("inMs"))?.toLong() ?: 0L
                    val outMs = (call.argument<Number>("outMs"))?.toLong() ?: 0L
                    val mode = call.argument<String>("mode") ?: "accurate"
                    if (path.isNullOrEmpty() || outPath.isNullOrEmpty()) {
                        result.error("ARG", "path+outPath required", null); return
                    }
                    trimmer.trim(path, inMs, outMs, outPath, mode, result)
                }
                "mix" -> {
                    val rawTracks = call.argument<List<Map<String, Any?>>>("tracks")
                    val outPath = call.argument<String>("outPath")
                    if (rawTracks.isNullOrEmpty() || outPath.isNullOrEmpty()) {
                        result.error("ARG", "tracks+outPath required", null); return
                    }
                    val targetLufs = (call.argument<Number>("targetLufs"))?.toDouble() ?: -14.0
                    val parsed = rawTracks.mapNotNull { m ->
                        val p = m["path"] as? String ?: return@mapNotNull null
                        val eq = m["eq"] as? Map<*, *>
                        AudioMixer.Track(
                            path = p,
                            inMs = (m["inMs"] as? Number)?.toLong() ?: 0L,
                            outMs = (m["outMs"] as? Number)?.toLong() ?: 0L,
                            positionMs = (m["positionMs"] as? Number)?.toLong() ?: 0L,
                            gainEnv = parseEnv(m["gainEnv"]),
                            panEnv = parseEnv(m["panEnv"]),
                            eqLowDb = (eq?.get("lowDb") as? Number)?.toDouble() ?: 0.0,
                            eqMidDb = (eq?.get("midDb") as? Number)?.toDouble() ?: 0.0,
                            eqHighDb = (eq?.get("highDb") as? Number)?.toDouble() ?: 0.0,
                            fadeInMs = (m["fadeInMs"] as? Number)?.toDouble() ?: 0.0,
                            fadeOutMs = (m["fadeOutMs"] as? Number)?.toDouble() ?: 0.0,
                            fadeKind = (m["fadeKind"] as? String) ?: "linear",
                            isVoiceForDuck = (m["isVoice"] as? Boolean) ?: false,
                        )
                    }
                    if (parsed.isEmpty()) {
                        result.error("ARG", "no valid tracks", null); return
                    }
                    mixer.mix(parsed, outPath, targetLufs, result)
                }
                "normalize" -> {
                    val path = call.argument<String>("path")
                    val outPath = call.argument<String>("outPath")
                    if (path.isNullOrEmpty() || outPath.isNullOrEmpty()) {
                        result.error("ARG", "path+outPath required", null); return
                    }
                    val targetLufs = (call.argument<Number>("targetLufs"))?.toDouble() ?: -14.0
                    normalizer.normalize(path, outPath, targetLufs, result)
                }
                "detectBeats" -> {
                    val path = call.argument<String>("path")
                    if (path.isNullOrEmpty()) {
                        result.error("ARG", "path required", null); return
                    }
                    beats.detect(path, result)
                }
                "pitchShift" -> {
                    val path = call.argument<String>("path")
                    val outPath = call.argument<String>("outPath")
                    val semitones = (call.argument<Number>("semitones"))?.toDouble() ?: 0.0
                    if (path.isNullOrEmpty() || outPath.isNullOrEmpty()) {
                        result.error("ARG", "path+outPath required", null); return
                    }
                    stretch.pitchShift(path, semitones, outPath, result)
                }
                "timeStretch" -> {
                    val path = call.argument<String>("path")
                    val outPath = call.argument<String>("outPath")
                    val factor = (call.argument<Number>("factor"))?.toDouble() ?: 1.0
                    val keepPitch = call.argument<Boolean>("keepPitch") ?: false
                    if (path.isNullOrEmpty() || outPath.isNullOrEmpty()) {
                        result.error("ARG", "path+outPath required", null); return
                    }
                    stretch.timeStretch(path, factor, keepPitch, outPath, result)
                }
                else -> result.notImplemented()
            }
        } catch (t: Throwable) {
            result.error("MINIS_AUDIO", t.message, null)
        }
    }

    private fun notImpl(result: MethodChannel.Result, method: String, anchor: String) {
        result.error(
            "NOT_IMPLEMENTED",
            "$method deferred (see improvement4.md $anchor)",
            null
        )
    }

    private fun parseEnv(raw: Any?): List<Pair<Double, Double>> {
        if (raw == null) return emptyList()
        val list = raw as? List<*> ?: return emptyList()
        val out = ArrayList<Pair<Double, Double>>(list.size)
        for (entry in list) {
            if (entry is List<*> && entry.size >= 2) {
                val t = (entry[0] as? Number)?.toDouble()
                val g = (entry[1] as? Number)?.toDouble()
                if (t != null && g != null) out.add(t to g)
            } else if (entry is Map<*, *>) {
                val t = (entry["timeMs"] as? Number)?.toDouble()
                val g = (entry["gain"] as? Number)?.toDouble()
                if (t != null && g != null) out.add(t to g)
            }
        }
        return out
    }

    companion object {
        @JvmStatic
        fun attach(binding: FlutterPlugin.FlutterPluginBinding): MinisAudioPlugin {
            return MinisAudioPlugin(
                binding.applicationContext,
                binding.binaryMessenger,
            )
        }
    }
}
